### ISSUE TITLE
Changes requested on extensions platform 

### ISSUE THREAD

**Issue by joepal1976** [makehumancommunity/mpfb2#231]
This is the first batch of comments from the extensions platform. More comments are expected later.

> Hello, welcome to the platform. This is quite a big add-on so probably will require more than 1 round of review.
> 
> I can't say this with certainty, but there seems to be some code mixed in which is for developing purposes and isn't exposed to users. It would be great if you could arrange those code in separate files so that they can be excluded from final .zip file. Anything that isn't needed for running the add-on is recommended to be left out. You can use Blenders command line arguments for building extension and exclude specific paths in blender_manifest to make this process easy.
> https://docs.blender.org/manual/en/latest/advanced/command_line/extension_arguments.html
> 
> This will make reviewing easier for us, because it's not always clear what is for users and needs to be reviewed and what is for developers that can be left alone.
> 
> I'm also generally not sure about how much of the code is actually used. Some stuff are named "legacy", also for example I wanted to see how export_to_disk function in makeskinmaterial.py worked but couldn't find it called anywhere. If add-on includes unused codes I would advise to clean-up.
> 
> Calls to potentially insecure functions exec() & eval()
> 
> In many cases getattr(data, attr) & setattr(data, attr, value) can be an alternative.
> In other cases this is used to load data from a file. Alternatives such as pickled data or JSON should be used instead.
> When a Python literal needs to be parse from an untrusted source consider using consider using ast.parse(...) then ast.literal_eval(node).
> I haven't fully bisected this because of code complexity, but I want to clarify that whenever you're using bpy.ops operators or properties from Rigify add-on (or add-on operators and functions that use them) first check if those properties exist and Rigify is on (best way to do this I think is checking for basic scene properties of Rigify). Otherwise it will throw nasty errors.
> 
> About internet permission "(optional) Interact with local instance of MakeHuman", can you point out where exactly that happens in code?

Some of these are more in the form of questions/confusion, and I'll answer those separately. But there are also things which need code changes. I'll hang those on this issue.

**UPDATE**: The second batch of comments have arrived

> Rigify checks aren't exactly correct.
> raise NotImplementedError('Rigify is not available. Download from extensions platform?')
> Rigify isn't on the platform, it's built-in add-on in core Blender (that's why depending on it is allowed for the time being). Also, I think this will only print message in console, not in UI as notification for users. Is there notifications for users so they can understand why something is failing? You need to use self.report('WARNING', "test") to make it appear in UI.
> 
> There is error when importing in rigifyhelpers.py:
> from rigify.utils.rig import resolve_layer_names
> I see that in locationservices.py you read and write to OS folders. Is there really a need for that? It adds unnecessary complexity and can freak out users a little bit. If I understand correctly it's just for storing add-on cache, and reading it, nothing external is read. In that case it's highly recommended to use bpy.utils.extension_path_user to create user-writable and safe directory exclusive to extension next to repositories and use that instead. It can really simplify things in code as well.
> 
> In lot of places, especially in righelpers object mode is switched very often. I'm not really familiar with armature code, but I think in most cases it might not be needed. If function only sets or reads property values it can be done in any more. Switch is necessary when calling bpy.ops operators or for other sensitive data that is constructed only when in specific mode.
> 
> 'Modeling' is probably not the correct tag for this.
> 
> When trying to create human I see no presets in dropdown, just empty enum, is there something missing?
> 
> Every operator/function that connects to web needs to include check for "Allow Online Access" user preference (bpy.app.online_access) before it executes. Best to include it in all polls.

**Comment by Vidyut** [comment]
I have no idea what this means, but I'm assuming that we're getting free code audits that might make MPFB even better :D

**Comment by joepal1976** [comment]
@Vidyut You're not wrong. :-) I don't think anyone else thoroughly read through and evaluated the code before (me included).

**Comment by joepal1976** [comment]
@Vidyut : If you happen to have time in between all the santas and reindeer and stuff, I could use some help trying out tomorrow's nightly build to see if something nasty has happened. 

The pertinent changes are:

- Better checks for rigify. Using a rigify rig or converting to a rigify rig should work as before when the rigify addon is enabled. But if rigify is disabled, sensible error messages should be shown when trying to use rigify rigs.
- Changed some background stuff in the model panel to avoid ugly code. Everything should look exactly as it did before (ie not crash unexpectedly)

If this works, I'll send it as an updated version to the extension platform and request another look at the addon.

**Comment by Vidyut** [comment]
Say when and I'll hit download.

**Comment by joepal1976** [comment]
Now the build is available.

Yes, trying to generate a few humans with rigify rigs and playing around with modling is basically the kind of testing that is needed. In the best of worlds there will be no incomprehensible error messages and the model panel should look and behave as before.

**Comment by Vidyut** [comment]
On it. Downloading.

**Comment by Vidyut** [comment]
>playing around with modling

Trying to make a male character. Unfortunately, I haven't made many male characters before, so I don't know if this is normal. Maybe as a short Indian woman, my idea of an "average" man is on the light side but this feels big/bulky for what I generated. I don't know if this is a modelling issue or my unfamiliarity with the males MPFB2 generates.
![image](https://github.com/user-attachments/assets/d5252b69-e42f-4232-b6ad-eafaf2fc8f33)
It is also offering breasts? And the options are size and not "none of the above"? Though the character doesn't have breasts.

My confusion apart, generated successfully. Played around with some controls. Seems fine.
![image](https://github.com/user-attachments/assets/eed9485d-5653-4525-9e95-3c612698f3cc)

**Comment by joepal1976** [comment]
Thanks, that was exactly what was needed.

Aparently I've missed a check at generate, since the pose bone thing should not even have been reached. But the suggestion to prevent this at UI level and hide the button or gray it out is better anyway.

**Comment by joepal1976** [comment]
And, yes... The male thingy is expected. There's really nothing fundamental in the model that prevents a male model with breasts, although it will look strange or often not be visible at all.

All sliders are a continuous scale, so a character can be 25% male and 75% female too. The create character from scratch with the distinct settings are just starting points, as "average" is just a point on the scale from -1 to +1.

The bulky male here is probably due to the "ideal proportions", but it does not seem strange at a quick glance.

**Comment by Vidyut** [comment]
Ah. Ideal proportions mean the "V-shape"? This should probably be mentioned in the vicinity of that selector. <-- suggestion

I probably need to redesign one female character. I've been struggling with her shoulders. She's supposed to be feminine, etc and the shoulders are just too... wonderwoman and her butt looks male. lol. I suspect this could be because I misunderstood what ideal proportions might look like on a woman if the slider works the same for male and female. ROFL.

But seriously, if this assumption is correct, mentioning the "V-shape" instead of ideal proportions might be useful - in fact, it could be a reason why people are even looking at other human generators. Women needing to look "sexy" is a thing in art and if their "ideal" is wonkifying them, they need to know instead of looking for other softwares. But even for males. If I want an athlete - he'll be in peak form, but won't look like he maxes out his gym subscription. "Ideal" is too open to interpretation. "V-shape" is more specific about what it does and is actually a reasonable feature, since it deviates quite a bit from the ordinary shape of both men and women.

**Comment by joepal1976** [comment]
Yeah, I agree. The labels should be updated before stable release. I'm thinking this stems from language that was established around 2010 and which wasn't checked with users.

**Comment by joepal1976** [comment]
@Vidyut : That's cute :-) I like the animations. Are they mixamo or manual?

**Comment by joepal1976** [comment]
Anyway, the second batch of comments have arrived. I haven't had any time to do anything about it yet. Apart from the mode switching thing I think the changes are pretty obvious. The mode switches need to be analyzed further.

**Comment by joepal1976** [comment]
@Vidyut (and everyone else)... As a head's up, I'm soon going to merge a code update which changes the default location of user files. If one has explicitly set the mpfb user data directory as per the "user data" section of https://static.makehumancommunity.org/mpfb/faq/important_config.html, then this change makes no difference. But if relying on the currently existing location, then the user files will need to be manually moved once the change comes into effect.

**Comment by Vidyut** [comment]
@joepal1976 the animations are a mix. Some from mixamo (body), Audio2face (lipsync), some manual. But even the mixamo ones are modified quite a bit, since they won't have the exact thing.

**Comment by Vidyut** [comment]
An alternative label for modeling could be "customizing"

**Comment by joepal1976** [comment]
> Add-on approved: "MPFB"

Yay.

I'll see to finishing the release notes and writing an announcement.

**Comment by Vidyut** [comment]
Cheers!

**Comment by joepal1976** [comment]
No particular problems reported on the extension platform. Closing this.
